### PR TITLE
fs: bring back macos-specific copyfile(3)

### PR DESCRIPTION
**Pull request by Artoria2e5**
copyfile(3) does not always do its own open according to permission rules. We can solve that by guarding it with one of our own opens. This way we can keep the cool clonefile thing for huge node_modules and still have sensible results.

MacOS does not have faccess(2) in case you are wondering why I did this. There is only access(2) with the "real UID" shenanigans that will bite us in the future.

This PR partially reverts #2233. Tests in the said PR are kept, and they should pass.

### CODE REVIEW

**Comment by Artoria2e5** [comment]
We need to do an `open` first per nodejs/node#26936. Amending.

File: src/unix/fs.c
Comment (bnoordhuis) [inline src/unix/fs.c]:
```suggestion
    return dstfd;
```
(it's already a libuv error code)

File: src/unix/fs.c
Comment (cjihrig) [inline src/unix/fs.c]: Can you move the `dst_flags` handling closer to where it's used.

File: src/unix/fs.c
Comment (cjihrig) [inline src/unix/fs.c]: I think this comment should explain a bit more (such as how `copyfile()` doesn't respect permissions) and possibly link to a GitHub issue.

**Comment by Artoria2e5** [comment]
fstat is not the best idea because ACL is a thing on macOS. The correct way to do so is via the access(2) family of APIs, but macOS does not have a faccess(2) that does EUID instead of the UID. A quick Google search I did yesterday finds CarbonStdCLib.o, apparently a C library polyfill, doing something similar for faccess.

Unlinking is what the OS X cp(1) command does for `-c` anyways, so it kind of is idiomatic in this use.

Gonna fix the branch thing later today (UTC+8).

**Comment by Artoria2e5** [comment]
Hmm, looks like there is faccessat(2) on macOS > 10.10. What's the OS support policy for libuv and nodejs?

**Comment by cjihrig** [comment]
libuv supports macOS >= 10.7. See the list of supported platforms [here](https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md).

**Comment by Artoria2e5** [comment]
A good-ish polyfill for our desired functionality is found in [gnulib euidaccess(3)](https://github.com/coreutils/gnulib/blob/master/lib/euidaccess.c). It gives two implementations: the correct but non-reentrant one using a setre{u,g}id-access-setre{u,g}id pattern and the traditional stat thing which is incorrect with ACL. 

Since faccessat(2) is a syscall, I can possibly get away with doing a version check like the clone part and syscall(2) the whole thing away. I still need to pick a way to polyfill this feature for [10.7, 10.10), so please do give your suggestion on that.

**Comment by richardlau** [comment]
I presume the gnulib implementation will be GPL and therefore cannot be used in libuv.

**Comment by Artoria2e5** [comment]
Yeah, someone will need to wipe their brain with only the phrase "implement an access check based on stat() or setuid and access(2)" left if they decided to implement it there.

**Comment by cjihrig** [comment]
With this change, the macOS implementation has more overlap with the Unix implementation than it previously did. The overlap is more prone to getting out of sync - for example, the Unix implementation no longer opens the destination with `O_TRUNC` like this PR does.

Maybe we should look into using `fcopyfile()` only for clone operations and use the common implementation for everything else. It would essentially be a macOS specific version of [this code block](https://github.com/libuv/libuv/blob/07e4168b67f6d06dac2c2e21c9a1976a0db9d312/src/unix/fs.c#L1216-L1229).

**Comment by Artoria2e5** [comment]
Makes sense. An `open()` is needed for the perm check anyways.

**Comment by Artoria2e5** [comment]
Meh, reasoning through the open fds and fallbacks with `clonefile(2)` is still too hairy for me.
* I need to `unlink` the file for `clonefile` to work, or it would fail with EEXIST. Yeah it takes paths, not fds.
* If `clonefile` does fail for some other reason, I would need to do the very verbose open thing again.

**Comment by arcanis** [comment]
Ping? Seems like this affects both Yarn and pnpm.

**Comment by bnoordhuis** [comment]
@Artoria2e5 Is this ready for re-review or are you still working things out?

**Comment by Artoria2e5** [comment]
There isn't much I can change right now, so please just throw more comments at me.

**Comment by cjihrig** [comment]
> Meh, reasoning through the open fds and fallbacks with clonefile(2) is still too hairy for me.

FWIW, I was suggesting to use `fcopyfile()` since it works with fds and doesn't require unlinking the file first.

**Comment by Artoria2e5** [comment]
That wouldn't be possible since fcopyfile() does not perform cloning. I am not sure if it's in the documentation, since copyfile.c is a bit easier to Google.

**Comment by Artoria2e5** [comment]
Oi…

**Comment by Artoria2e5** [comment]
I don't think there is anything more I can do.

**Comment by arcanis** [comment]
File cloning is very important to package managers. This makes this PR very important for a lot of people, especially when a good chunk of the JS developer ecosystem uses OSX on their day-to-day work 😕

**Comment by wojtekmaj** [comment]
Bump. I refuse to let this PR die.

**Comment by cjihrig** [comment]
> That wouldn't be possible since fcopyfile() does not perform cloning.

FWIW, I tested `fcopyfile()` with the `COPYFILE_CLONE_FORCE`, and it appeared to work (at least it didn't return an error):

```diff
diff --git a/src/unix/fs.c b/src/unix/fs.c
index 88ac0731..6186289a 100644
--- a/src/unix/fs.c
+++ b/src/unix/fs.c
@@ -62,6 +62,7 @@
 #endif
 
 #if defined(__APPLE__)
+# include <copyfile.h>
 # include <sys/sysctl.h>
 #elif defined(__linux__) && !defined(FICLONE)
 # include <sys/ioctl.h>
@@ -1259,6 +1260,37 @@ static ssize_t uv__fs_copyfile(uv_fs_t* req) {
       goto out;
     }
   }
+#elif defined(__APPLE__) && !TARGET_OS_IPHONE
+  if (req->flags & UV_FS_COPYFILE_FICLONE ||
+      req->flags & UV_FS_COPYFILE_FICLONE_FORCE) {
+    static int can_clone;
+    copyfile_flags_t flags;
+    char buf[64];
+    size_t len;
+    int major;
+
+    /* Check OS version. Cloning is only supported on macOS >= 10.12. */
+    if (req->flags & UV_FS_COPYFILE_FICLONE_FORCE) {
+      if (can_clone == 0) {
+        len = sizeof(buf);
+        if (sysctlbyname("kern.osrelease", buf, &len, NULL, 0))
+          return UV__ERR(errno);
+
+        if (1 != sscanf(buf, "%d", &major))
+          abort();
+
+        can_clone = -1 + 2 * (major >= 16);  /* macOS >= 10.12 */
+      }
+
+      if (can_clone < 0)
+        return UV_ENOSYS;
+    }
+
+    flags = COPYFILE_ALL | (1 << 25 /* COPYFILE_CLONE_FORCE */);
+    err = fcopyfile(srcfd, dstfd, NULL, flags);
+    if (err == 0 || req->flags & UV_FS_COPYFILE_FICLONE_FORCE)
+      goto out;
+  }
 #else
   if (req->flags & UV_FS_COPYFILE_FICLONE_FORCE) {
     err = UV_ENOSYS;
```

**Comment by cjihrig** [comment]
I may have spoken too soon:

```
     COPYFILE_CLONE_FORCE   Clone the file instead.  This is a force flag i.e.
                            if cloning fails, an error is returned.  This flag
                            is equivalent to (COPYFILE_EXCL | COPYFILE_ACL |
                            COPYFILE_STAT | COPYFILE_XATTR | COPYFILE_DATA |
                            COPYFILE_NOFOLLOW_SRC).  Note that if cloning is
                            successful, progress callbacks will not be
                            invoked.  Note also that there is no support for
                            cloning directories: if a directory is provided as
                            the source, an error will be returned.  (This is
                            only applicable for the copyfile() function.)
```

**Comment by arcanis** [comment]
Ping to prevent staleclose?

**Comment by zkochan** [comment]
ping

**Comment by noahnu** [comment]
ping

**Comment by wojtekmaj** [comment]
Ping 🏓

**Comment by wojtekmaj** [comment]
Bad bot

**Comment by mischnic** [comment]
ping

**Comment by bnoordhuis** [comment]
Do I summarize correctly this pull request is blocked indefinitely as long as libuv supports old macos versions?

Personally, I'm okay with raising the minimum version (10.7 is 10 years old) but, on the other hand, I know there are still a handful of users left.

**Comment by wojtekmaj** [comment]
Well it's not like they couldn't just... Use the old version. It's what 10.7 users are very much into anyway.

**Comment by zkochan** [comment]
What value does this stale bot bring to the conversation?

**Comment by arcanis** [comment]
It's really quite frustrating indeed. @bnoordhuis @cjihrig what does the OP need to do to bring this PR to the finish line?

**Comment by gengjiawen** [comment]
@bnoordhuis @cjihrig Anything would be a blocker here ?

**Comment by zkochan** [comment]
ping

**Comment by Jarred-Sumner** [comment]
To give you a sense of the impact here — bun’s implementation of fs.copyFileSync which uses copyfile() on macOS performs 2x faster than node’s fs.copyFileSync when copying a 1 GB file.

59392F5C-1225-4D3E-9A9E-7D8E1FFDF571

**Comment by bnoordhuis** [comment]
I opened #3406 earlier this week to raise the baseline to macOS 10.15. When accepted and merged, I think that would unblock this PR.

**Comment by uvesten** [comment]
> To give you a sense of the impact here — bun’s implementation of fs.copyFileSync which uses copyfile() on macOS performs 2x faster than node’s fs.copyFileSync when copying a 1 GB file.

That's good, but I thought the main issue here was that APFS Copy on Write is not supported. When it is supported, the speedup should be closer to infinity than 2x.

**Comment by Jarred-Sumner** [comment]
This wasn’t with the COPYFILE_FICLONE flag, so whether or not it used copy
on write is up to the implementation of copyfile()

**Comment by bnoordhuis** [comment]
The baseline is now macos 10.15. There's one remaining comment from @cjihrig so I defer to him on whether it's ready for merging. Let me know if I should also review.

**Review by cjihrig — APPROVED** [review]
LGTM if the tests pass.

File: src/unix/fs.c
Comment (bnoordhuis) [inline src/unix/fs.c:1173]: With 10.15 as the baseline, it's presumably okay to use faccessat() now?

File: src/unix/fs.c
Comment (bnoordhuis) [inline src/unix/fs.c:1190]: Stylistic:
```suggestion
  if (dstfd < 0)
    return dstfd;

  err = uv__close_nocheckstdio(dstfd);
  if (err)
    return err;

  /* We might have created it. We also don't want clone to fail without
   * UV_FS_COPYFILE_EXCL. */
  unlink(req->new_path);

  if (copyfile(req->path, req->new_path, NULL, flags))
    return UV__ERR(errno);

  return 0;
```

File: src/unix/fs.c
Comment (bnoordhuis) [inline src/unix/fs.c:1191]:
```suggestion
#else  /* defined(__APPLE__) && !TARGET_OS_IPHONE */
```

**Comment by zkochan** [comment]
Never Gonna Give You Up

**Comment by FezVrasta** [comment]
Adding my comment so my anti stale bot will take care of this pr from now on

**Comment by bnoordhuis** [comment]
There's some unaddressed feedback I left in February. Most is stylistic but the faccessat() question needs an answer before this can go in.

Also, a gentle reminder that bots and their owners are likely to meet with swift and permanent bans.

**Comment by zkochan** [comment]
Where is the anti stale bot?

**Comment by cjihrig** [comment]
I've added the not-stale label. Hopefully that will keep stale bot away.

**Comment by bnoordhuis** [comment]
@Artoria2e5 Do you plan on picking up this pull request again? If not, maybe someone else wants to step up and get this over the finish line?

**Comment by mischnic** [comment]
> If not, maybe someone else wants to step up and get this over the finish line?

I've tried to incorporate your feedback: https://github.com/libuv/libuv/pull/3654

**Comment by Artoria2e5** [comment]
In light of an updated PR being available I am closing this one. Just don't have time for code when I am all stuck at McD, I am afraid.

**Pull request closed without merging**
